File: apps/backend/app/api/v1/note.py
"""
笔记相关路由。
"""

from __future__ import annotations

from fastapi import APIRouter, Query

from app.api.deps import CurrentUser, DbSession, LargePagination
from app.schemas.note import NoteCreate, NoteListResponse, NoteResponse, NoteUpdate
from app.schemas.user import MessageResponse
from app.services.note_service import note_service

router = APIRouter(prefix="/notes", tags=["笔记"])


@router.post("", response_model=NoteResponse, summary="创建笔记")
async def create_note(
    note_in: NoteCreate,
    db: DbSession,
    current_user: CurrentUser,
) -> NoteResponse:
    note = await note_service.create(db=db, user_id=current_user.id, obj=note_in)
    return NoteResponse.model_validate(note)


@router.get("", response_model=NoteListResponse, summary="获取笔记列表")
async def get_notes(
    pagination: LargePagination,
    db: DbSession,
    current_user: CurrentUser,
    article_id: int | None = Query(None, description="按文章筛选"),
) -> NoteListResponse:
    return await note_service.get_list(
        db=db,
        user_id=current_user.id,
        page=pagination.page,
        page_size=pagination.page_size,
        article_id=article_id,
    )


@router.patch("/{note_id}", response_model=NoteResponse, summary="更新笔记")
async def update_note(
    note_id: int,
    note_in: NoteUpdate,
    db: DbSession,
    current_user: CurrentUser,
) -> NoteResponse:
    note = await note_service.update(
        db=db,
        user_id=current_user.id,
        note_id=note_id,
        obj=note_in,
    )
    return NoteResponse.model_validate(note)


@router.delete("/{note_id}", response_model=MessageResponse, summary="删除笔记")
async def delete_note(
    note_id: int,
    db: DbSession,
    current_user: CurrentUser,
) -> MessageResponse:
    await note_service.delete(db=db, user_id=current_user.id, note_id=note_id)
    return MessageResponse(message="笔记删除成功")

File: apps/web/src/api/note.ts
/**
 * 笔记模块 API
 *
 * 提供笔记的增删改查功能，支持关联文章创建笔记
 */
import request from '@/utils/request';

/**
 * 笔记数据模型
 */
export interface Note {
  /** 笔记 ID */
  id: number;
  /** 用户 ID */
  user_id: number;
  /** 笔记标题（可选） */
  title?: string;
  /** 笔记内容 */
  content: string;
  /** 关联文章 ID（可选） */
  article_id?: number;
  /** 关联文章标题（可选，后端返回） */
  article_title?: string;
  /** 创建时间 */
  created_at: string;
  /** 更新时间 */
  updated_at: string;
}

/**
 * 创建笔记请求参数
 */
export interface CreateNoteRequest {
  /** 笔记标题（可选） */
  title?: string;
  /** 笔记内容（必填） */
  content: string;
  /** 关联文章 ID（可选） */
  article_id?: number;
}

/**
 * 更新笔记请求参数
 */
export interface UpdateNoteRequest {
  /** 笔记标题（可选） */
  title?: string;
  /** 笔记内容（可选） */
  content?: string;
}

/**
 * 笔记列表响应数据
 */
export interface NoteListResponse {
  /** 笔记列表 */
  items: Note[];
  /** 总数 */
  total: number;
}

/**
 * 创建笔记
 * @param data - 创建笔记的请求数据
 * @returns 创建成功的笔记对象
 */
export async function createNote(data: CreateNoteRequest): Promise<Note> {
  return request.post('/notes', data);
}

/**
 * 获取笔记列表
 * @param params - 查询参数
 * @param params.page - 页码（可选，默认 1）
 * @param params.page_size - 每页数量（可选）
 * @param params.article_id - 按文章 ID 筛选（可选）
 * @returns 笔记列表响应数据
 */
export async function getNotes(params: {
  page?: number;
  page_size?: number;
  article_id?: number;
}): Promise<NoteListResponse> {
  return request.get('/notes', { params });
}

/**
 * 更新笔记
 * @param id - 笔记 ID
 * @param data - 更新笔记的请求数据
 * @returns 更新后的笔记对象
 */
export async function updateNote(id: number, data: UpdateNoteRequest): Promise<Note> {
  return request.patch(`/notes/${id}`, data);
}

/**
 * 删除笔记
 * @param id - 笔记 ID
 * @returns 删除结果
 */
export async function deleteNote(id: number): Promise<{ success: boolean }> {
  return request.delete(`/notes/${id}`);
}
